### ISSUE TITLE
Failing merge for text values

### ISSUE THREAD

**Issue by cetra3** [automerge/automerge#528]
This might be a misunderstanding on my part, but the follow test fails spuriously:

```rust
// Initialise site_1 with blank text
let mut site_1 = AutoCommit::new();
let site_1_text_id = site_1.put_object(ROOT, "text", ObjType::Text).unwrap();
site_1.splice_text(&site_1_text_id, 0, 0, "").unwrap();

// Initialise site_2 & add `testing automerge` to the buffer
let mut site_2 = AutoCommit::new();
let site2_text_id = site_2.put_object(ROOT, "text", ObjType::Text).unwrap();
site_2.splice_text(&site2_text_id, 0, 0, "testing automerge").unwrap();

// Save site_2 & merge into site_1
let state = site_2.save();
let mut site_2_load = AutoCommit::load(&state).unwrap();
site_1.merge(&mut site_2_load).unwrap();

// Get the value at `text` and assert it's equal to `testing automerge`
let text_id = site_1.get(ROOT, "text").unwrap().unwrap().1;
assert_eq!(site_1.text(text_id).unwrap(), "test")
```

**Comment by alexjg** [comment]
There are two issues here:

1. You are using `Transaction::set` instead of `Transaction::spllice_text`. This would result in a conflict on the `"text"` key.
2. More fundamentally, you start from an empty document rather than initialising the document and then passing it to the concurrent nodes.

This second point is a little tricky to understand and people fall foul of it all the time, we need to find a better API for it. Basically  the problem is that automerge is more like a git repository than a JSON document. If you tried to merge two independent git repositories with the same value for a given file you would end up with merge conflicts on that file. On the other hand, if you check out an initial version of the repository and then make a change to the file and merge it back with the initial repository you don't have merge conflicts. Automerge is similar, in order to know that two changes are related to the same object you need to have the same object ID, this means for the usecase you're talking about here: you would first create a document containing an empty text value, then distribute that to your concurrent peers (analogous to cloning a git repo), then make your concurrent changes.

For example:

```rust
use automerge::{transaction::Transactable, ReadDoc};

fn main() {
    let mut doc = automerge::AutoCommit::new();
    let text_id = doc.put_object(automerge::ROOT, "text", automerge::ObjType::Text).unwrap();

    let mut doc2 = doc.fork();
    doc2.splice_text(&text_id, 0, 0, "hello").unwrap();

    doc.splice_text(&text_id, 0, 0, " world").unwrap();
    doc.merge(&mut doc2).unwrap();

    println!("{}", doc.text(&text_id).unwrap());
}
```

This will either print `"hello world"` or `" worldhello"` depending on what actor IDs are chosen.

**Comment by cetra3** [comment]
Hey thanks for the response!

> You are using Transaction::set instead of Transaction::spllice_text. This would result in a conflict on the "text" key.

I'm using `splice_text` in this example, I'm not sure where I am using `set`? Unless there is a method that calls it.

> More fundamentally, you start from an empty document rather than initialising the document and then passing it to the concurrent nodes.

This is a hard one, because intuitively I would think that CRDTs don't need to be from the same starting document?  Like if you have a grow-only set, they don't need to be derived from a single initialised state. I have used other CRDTs libraries in the past that don't have this constraint, let me know if this is a limitation with automerge.

Because now you have the challenge of if you have completely independent sites/actors create state and work on a document independently, and then some time in the future, they connect & synchronise it, who wins?

**Comment by ept** [comment]
> This is a hard one, because intuitively I would think that CRDTs don't need to be from the same starting document? Like if you have a grow-only set, they don't need to be derived from a single initialised state. I have used other CRDTs libraries in the past that don't have this constraint, let me know if this is a limitation with automerge.

You can merge two Automerge docs that were created independently, and any keys you set on the root object will merge in the way you expect. However, if you create nested objects (in your case, the Text object under the key `text`), then each nested object has its own identity in Automerge. You were creating two separate Text objects under the same key, and in this case Automerge records a conflict on the top-level key, and keeps the two Text objects as separate objects.

This is a long-standing problem; see issue #4 for a discussion (dating back to the days when Automerge was still called Tesseract). Given that we keep running into the same problem, it would be good to finally work on solving it. I can see two potential solutions:

1. When two nested objects of the same type are concurrently created under the same key, we merge them into one object (interpret ops on both objects as if they had been applied to the same object, e.g. the one with the lower objectId, and don't record a conflict on the key).
2. Change the API so that object creation has to happen through a separate "schema change" mechanism (which could then incorporate concepts from [Cambria](https://www.inkandswitch.com/cambria/)).

**Comment by alexjg** [comment]
@cetra3 what other libraries have you used that don't have this problem? All the libraries I have used do have this issue with nested objects (e.g. in Yjs if you create an array inside a map on two nodes and merge them then the result will be only one of the maps being visible).


@ept  I feel like `1` may be confusing with nested sequences. Imagine we had two concurrent documents like this:

```
{
    contacts: [
        {
            name: "Alice",
            email: "alice@example.org"
        }   
    ]
}
```

```
{
    contacts: [
        {
            name: "Bob",
            email: "bob@example.org"
        }   
    ]
}
```

If we were to merge nested objects of the same type by key we would end up with conflicts on the `name` and `email` keys right? (Because we would first merge the two lists under the `"contacts"` key, then the two maps under the `0` key)? Or would this only apply to map keys (I possible am thinking of an overly general use of the term "key")? Does this play nice with the (as yet unimplemented) move operation? I think it does.

The schema change idea sounds tantalizing!

**Comment by cetra3** [comment]
>You can merge two Automerge docs that were created independently, and any keys you set on the root object will merge in the way you expect

I couldn't actually see any way of setting keys on the root object itself?  I.e, it looks like the root object itself is hard-set to a map. I tried a `splice_text` on the `ROOT` but it returned an error.

> what other libraries have you used that don't have this problem? All the libraries I have used do have this issue with nested objects (e.g. in Yjs if you create an array inside a map on two nodes and merge them then the result will be only one of the maps being visible).

I was previously using [ditto](https://github.com/alex-shapiro/ditto), but that now looks unmaintained.  This wasn't a full blown "object" though, I was just using the text CRDT directly, which I couldn't see any functions/behaviour to do that with automerge (I.e, don't have an object, just text).

**Comment by ept** [comment]
@alexjg List insertions would remain independent from each other, because they have different list element IDs, and we should use element IDs (not indexes) as keys. In your example, the merge would produce a single contacts list containing two objects, one with Alice's details and one with Bob's details. The discussion in #4 also seemed to favour this behaviour. Regarding a move operation, merging objects under the same key would introduce a bit of extra complexity (because we would have to treat the object IDs of all the merged objects as aliases for each other), but I can't think of any serious problem.

**Comment by ept** [comment]
> I couldn't actually see any way of setting keys on the root object itself? I.e, it looks like the root object itself is hard-set to a map. I tried a splice_text on the ROOT but it returned an error.

@cetra3 Correct, the root object is always a map. You are already setting keys on the root object by calling `.put_object(ROOT, "text", ...)`
